framework/src/media/MediaPlayerImpl.cpp

```diff
@@ -146,96 +222,194 @@ player_result_t MediaPlayerImpl::unprepare()
 
 	mCurState = PLAYER_STATE_IDLE;
-	return PLAYER_OK;
+	ret = PLAYER_OK;
+exit:
+	mSyncCv.notify_one();
 }
 
 player_result_t MediaPlayerImpl::start()
 {
 	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer start\n");
-	if (mCurState == PLAYER_STATE_NONE || mCurState == PLAYER_STATE_IDLE) {
-		meddbg("MediaPlayer unprepare fail : wrong state\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
 		return PLAYER_ERROR;
 	}
 
-	PlayerWorker& mpw = PlayerWorker::getWorker();
-	mpw.getQueue().enQueue(&PlayerWorker::startPlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::startPlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::startPlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : startPlayer\n");
+	if (mCurState != PLAYER_STATE_READY && mCurState != PLAYER_STATE_PAUSED) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	if (mpw.getPlayer() != nullptr) {
+		mpw.getPlayer()->pausePlayer();
+	}
+	mpw.setPlayer(shared_from_this());
+
+	mCurState = PLAYER_STATE_PLAYING;
+	notifyObserver(PLAYER_OBSERVER_COMMAND_STARTED);
+}
+
 player_result_t MediaPlayerImpl::stop()
 {
 	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer stop\n");
-	if (mCurState == PLAYER_STATE_NONE || mCurState == PLAYER_STATE_IDLE) {
-		meddbg("MediaPlayer stop fail : wrong state\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
 		return PLAYER_ERROR;
 	}
 
-	PlayerWorker& mpw = PlayerWorker::getWorker();
-	mpw.getQueue().enQueue(&PlayerWorker::stopPlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::stopPlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::stopPlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : stopPlayer\n");
+	if (mCurState != PLAYER_STATE_PLAYING && mCurState != PLAYER_STATE_PAUSED) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	mpw.setPlayer(nullptr);
+	mCurState = PLAYER_STATE_READY;
+}
+
 player_result_t MediaPlayerImpl::pause()
 {
 	std::lock_guard<std::mutex> lock(mCmdMtx);
 	medvdbg("MediaPlayer pause\n");
-	if (mCurState == PLAYER_STATE_NONE || mCurState == PLAYER_STATE_IDLE) {
-		meddbg("MediaPlayer pause fail : wrong state\n");
+
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+	if (!mpw.isAlive()) {
 		return PLAYER_ERROR;
 	}
 
-	PlayerWorker& mpw = PlayerWorker::getWorker();
-	mpw.getQueue().enQueue(&PlayerWorker::pausePlayer, &mpw, shared_from_this());
+	mpw.getQueue().enQueue(&MediaPlayerImpl::pausePlayer, shared_from_this());
 
 	return PLAYER_OK;
 }
 
+void MediaPlayerImpl::pausePlayer()
+{
+	PlayerWorker& mpw = PlayerWorker::getWorker();
+
+	medvdbg("MediaPlayer Worker : pausePlayer\n");
+	if (mCurState != PLAYER_STATE_PLAYING) {
+		notifyObserver(PLAYER_OBSERVER_COMMAND_ERROR);
+		return;
+	}
+
+	mpw.setPlayer(nullptr);
```

Comment:
Currently start calls setPlayer(this) and pause calls setPlayer(nullptr), it means every pcm_readi/writei operation, it getting player, so This must be considerated, because it is overload(We should reduce delay of start!!)
If application paused by user, Definetely we don't have to set nullptr when pause.
Also I think if player paused by other application, flow is below
App A. preapare -> start -> paused(maybe callback called by App B)
App B. request audio (App a paused) -> prepare -> start -> stop 
When app B stopped, Audio gain callback will come, then it will call prepare -> start again.
So, Every time when it change player, it should call prepare before start.
How do you think?